applications/IgaApplication/custom_elements/shell_5p_hierarchic_element.cpp

```diff
@@ -0,0 +1,1307 @@
+        rInitialTransCarToCov(4, 3) = rInitialTransCarToCov(4, 3) / 2.0;
+        rInitialTransCarToCov(4, 4) = rInitialTransCarToCov(4, 4) / 2.0;
+    }
+
+    void Shell5pHierarchicElement::CalculateTransformationMatrixActualTransCovToCar(
+        Matrix& rActualTransCovToCar,
+        const Vector& rg1,
+        const Vector& rg2,
+        const Vector& rg3,
+        const Vector& ra2_con,
+        const Vector& ra3_kirchhoff_love)
+    {
+        // local cartesian coordinates, direction of a2_con and g2_con is the same -> it is possible to use ra2_con instead of g2_con 
+        array_1d<double, 3> e1 = rg1 / norm_2(rg1);
+        array_1d<double, 3> e2 = ra2_con / norm_2(ra2_con);
+        // e3 = a3_kirchhoff_love
+
+        double G_00 = inner_prod(e1, rg1);
+        double G_01 = inner_prod(e1, rg2);
+        double G_02 = inner_prod(e1, rg3);
+        double G_11 = inner_prod(e2, rg2);
+        double G_12 = inner_prod(e2, rg3);
+        double G_22 = inner_prod(ra3_kirchhoff_love, rg3);
+ 
+        rActualTransCovToCar(0, 0) = pow(G_00, 2);
+        rActualTransCovToCar(0, 1) = pow(G_01, 2);
+        rActualTransCovToCar(0, 2) = 2 * G_00 * G_01;
+        rActualTransCovToCar(0, 3) = 2 * G_01 * G_02;
+        rActualTransCovToCar(0, 4) = 2 * G_00 * G_02;
+        rActualTransCovToCar(1, 1) = pow(G_11, 2);
+        rActualTransCovToCar(1, 3) = 2 * G_11 * G_12;
+        rActualTransCovToCar(2, 1) = G_01 * G_11;
+        rActualTransCovToCar(2, 2) = G_00 * G_11;
+        rActualTransCovToCar(2, 3) = G_01 * G_12 + G_02 * G_11;
+        rActualTransCovToCar(2, 4) = G_00 * G_12;
+        rActualTransCovToCar(3, 3) = G_11;
+        rActualTransCovToCar(4, 3) = G_01 * G_22;
+        rActualTransCovToCar(4, 4) = G_00 * G_22;
+    }
+
+    void Shell5pHierarchicElement::CalculateShearDifferenceVector(
+        array_1d<double, 3>& rw,
+        array_1d<double, 3>& rDw_D1,
+        array_1d<double, 3>& rDw_D2,
+        array_1d<double, 2>& rw_alpha,
+        Matrix& rDw_alpha_Dbeta,
+        const MetricVariables& rActualMetric,
+        IndexType IntegrationPointIndex)
+    {
+        const GeometryType& r_geometry = GetGeometry();
+        const Matrix& r_N = r_geometry.ShapeFunctionsValues();
+        const Matrix& r_DN_De = r_geometry.ShapeFunctionDerivatives(1, IntegrationPointIndex);
+        const SizeType number_of_control_points = r_geometry.size();
+        double w_1, w_2;
+
+        for (IndexType i = 0; i < number_of_control_points; ++i) 
+        {
+            // only ROTATION_X and ROTATION_Y used preliminarily, to avoid new declarations
+            // ROTATION_X = w_1 (first component of hierarchic shear difference vector)
+            // ROTATION_Y = w_2 (second component of hierarchic shear difference vector)
+            w_1 = r_geometry[i].GetDof(ROTATION_X).GetSolutionStepValue();
+            w_2 = r_geometry[i].GetDof(ROTATION_Y).GetSolutionStepValue();

```

Comment:
Tih would also benefit from #7385. Since the more resonable variable names can be simply defined